packages/astro/src/vite-plugin-astro-server/index.ts

```diff
@@ -79,37 +89,64 @@ async function handleRequest(
 	const origin = `${viteServer.config.server.https ? 'https' : 'http'}://${req.headers.host}`;
 	const pathname = decodeURI(new URL(origin + req.url).pathname);
 	const rootRelativeUrl = pathname.substring(devRoot.length - 1);
+
 	try {
 		if (!pathname.startsWith(devRoot)) {
-			info(logging, 'astro', msg.req({ url: pathname, statusCode: 404 }));
+			log404(logging, pathname);
 			return handle404Response(origin, config, req, res);
 		}
 		// Attempt to match the URL to a valid page route.
 		// If that fails, switch the response to a 404 response.
 		let route = matchRoute(rootRelativeUrl, manifest);
 		const statusCode = route ? 200 : 404;
-		// If no match found, lookup a custom 404 page to render, if one exists.
+
 		if (!route) {
-			const relPages = config.pages.href.replace(config.projectRoot.href, '');
-			route = manifest.routes.find((r) => r.component === relPages + '404.astro');
+			log404(logging, pathname);
+			const custom404 = getCustom404Route(config, manifest);
+			if (custom404) {
+				route = custom404;
+			} else {
+				return handle404Response(origin, config, req, res);
+			}
 		}
-		// If still no match is found, respond with a generic 404 page.
-		if (!route) {
-			info(logging, 'astro', msg.req({ url: pathname, statusCode: 404 }));
-			handle404Response(origin, config, req, res);
-			return;
+
+		// Note: may re-assign filePath to custom 404 on error
+		let filePath = new URL(`./${route.component}`, config.projectRoot);
+		// Note: may re-assign preloaded component to custom 404 on error
+		let preloadedComponent = await preload({ astroConfig: config, filePath, viteServer });
```

Comment:
I don't _love_ the idea of reassigning these, but I wouldn't block this PR for that reason. Is there a cleaner way to do this, though?

Reply:
I thought about this for _way_ too long 😆 I agree, reassigning isn't very FP of me. I'll give a non-reassigning solution a try here!